### PR TITLE
CRAYSAT-1410: Allow session templates to refer to images by ref_name

### PR DESCRIPTION
**Pull request by haasken-hpe**
## Summary and Scope

Add functionality to `sat bootprep` to allow session templates in the
input file to refer to images by their `ref_name` value. This is a
usability improvement to make it easier to refer to an image which has a
name that uses its base image or recipe name in its name.

Deprecate the old way of specifying the IMS image name or ID as a string
in the session template. Add the ability to specify an object under the
`image` property that has either an `ims` property or an `image_ref`
property. The value of the `ims` property is a dict with either a `name`
or an `id` property. The value of the `image_ref` property is a string
that refers to an image in the input file by the value of its `ref_name`
property.

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-1410](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1410)

### Tested on:

  * vale
  * python virtual env on macbook (generate-docs)

### Test description:
Tested on vale with a bootprep input file that creates session templates
which specify their images using `image.ims.id`, `image.ims.name`, or
`image.image_ref`.

Also tested creating a session template with `image` set to a string
containing either the `id` or the `name` of an image.

## Risks and Mitigations

None

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Review by ekamin-hpe — CHANGES_REQUESTED** [review]
I think we should address the edge case I brought up. Other than that, looks good.

File: sat/cli/bootprep/input/session_template.py
Comment (ekamin-hpe) [reply sat/cli/bootprep/input/session_template.py:169]: Addressed with c9ae744a7f4666cd5330a09162ddd416d655b1dd

**Review by ekamin-hpe — APPROVED** [review]
I pushed a commit addressing my concern and am now approving, although we would like to complete the review of #35 before merging this, so that it is easier to keep the commit history clear.

**Review by jack-stanek-hpe — APPROVED** [review]
Looks good to me, outside of one bikeshed comment. I don't think it's worth holding up merging though.

File: sat/cli/bootprep/input/session_template.py
Comment (haasken-hpe) [reply sat/cli/bootprep/input/session_template.py:175]: I agree. I was thinking about this while writing the code, but when I started to try to put more meaning into the names, it did feel like they were getting long and hard to parse. Your suggestions are pretty good though. I'll consider it further.

File: sat/cli/bootprep/input/session_template.py
Comment (haasken-hpe) [reply sat/cli/bootprep/input/session_template.py:169]: I discussed this with Eli this morning. Here's a summary of our discussion.

This issue only happens during dry run. Here is an example of running `sat bootprep run --dry-run` against Eli's third example:

```
ncn-m001:~/haasken/bootprep # sat bootprep run --dry-run eli-example3.yaml
INFO: Validating given input file eli-example3.yaml
INFO: Input file successfully validated against schema
INFO: Given input did not define any CFS configurations
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: Found IMS base for image at index 0: recipe with name=cray-shasta-csm-sles15sp2-barebones.x86_64-shasta-1.5
INFO: Of the 1 that would be created, 1 have no dependencies and would be created first.
INFO: Would create 1 images.
INFO: Dry run, not creating images.
ERROR: No image with name elis_image exists for use by BOS session template at index 0.
ERROR: The BOS session template at index 0 is not valid. See errors above.
ERROR: One or more items is not valid in collection of BOS session templates
ERROR: The collection of BOS session templates is not valid. See errors above.
```

Note that this error does not occur if we are not in dry-run mode. The session template validation occurs after `sat.cli.bootprep.image.create_images` is called. When not in dry-run mode, by the time we get to session template validation, the image will have actually been created, and so the session template will validate. E.g.:

```
ncn-m001:~/haasken/bootprep # sat bootprep run  eli-example3.yaml
INFO: Validating given input file eli-example3.yaml
INFO: Input file successfully validated against schema
INFO: Given input did not define any CFS configurations
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: Found IMS base for image at index 0: recipe with name=cray-shasta-csm-sles15sp2-barebones.x86_64-shasta-1.5
INFO: Of the 1 that will be created, 1 have no dependencies and will be created first.
INFO: Creating 1 images.
INFO: Creating images
INFO: Launching IMS job to create image
INFO: Created IMS image creation job with ID 743e6f56-9b01-4804-890d-2d01c1c070c0
INFO: Creation of image elis_image succeeded: ID 04eda969-f279-48b6-b934-ddfd3267b8bf
INFO: Image elis_image does not need configuration.
INFO: Creation of image elis_image succeeded: ID 04eda969-f279-48b6-b934-ddfd3267b8bf
INFO: Image creation completed successfully
INFO: Creating 1 BOS session template
```

After talking with Eli, I still think it's a valid concern that we should continue to allow session template validation to work in dry-run mode, and I think Eli's fix is correct. I'm going to test a little bit further with that fix. My main concern would be whether it's impacted at all by the possibility that image names can now be rendered using the base image or recipe. I think it will be fine, but I'm going to test that.

File: sat/cli/bootprep/input/session_template.py
Comment (haasken-hpe) [reply sat/cli/bootprep/input/session_template.py:180]: I'm going to ignore this for now since we're on a tight schedule, but I'm definitely open to a rename in the future. It doesn't affect anything user-facing, so it can be changed easily later.

**Comment by haasken-hpe** [comment]
> I pushed a commit addressing my concern and am now approving, although we would like to complete the review of #35 before merging this, so that it is easier to keep the commit history clear.

Thanks! I squashed that commit into mine before doing a rebase on feature/CRAYSAT-1385 because it made resolving conflicts a little easier for me. I made some slight changes to the approach, summarized here:

* Incorporated the `provides_context('image')` decorator on the `validate_image_exists` method as in the PR #35 for CRAYSAT-1411.
* Enforced requirement that `image.image_ref` in the session template resolves to an image in the input instance. Raise a validation error if it does not.
* Refactored the logic for looking up the input image based on the`image.image_ref` into a separate cached property since it's done in both `validate_image_exists` and the `image_record` property.

**Comment by haasken-hpe** [comment]
I'm going to be testing this now as described in the commit's test description. I will report how that goes in another comment.

File: sat/cli/bootprep/input/session_template.py
Comment (jack-stanek-hpe) [reply sat/cli/bootprep/input/session_template.py:180]: Sounds good. It definitely isn't super high priority at the moment.

**Comment by haasken-hpe** [comment]
Testing on vale went well. I tested by running `sat bootprep run` against the input files here: https://github.hpe.com/ryan-haasken/sprint-reviews/tree/main/2022-08-02/CRAYSAT-1410

I ran against each file both with and without `--dry-run` to ensure the session template validation works in both the dry-run and non-dry-run case.

**Pull request merged**
